### DISCUSSION TITLE
Any way to disable 'Skipping item - invalid name (Microsoft Naming Convention)' warnings for .lock?

### DISCUSSION THREAD

**Discussion by gusruben** [abraunegg/onedrive#2160] [Q&A] [answered]
I'm having an issue where my terminal is being clogged with these warnings every time I synchronize. There are lots of .lock files in my OneDrive, and I get a warning for every single one. I'm fine with the fact that they aren't being synchronized, but is there a way to disable specifically this warning?

I have tried blacklisting `.lock` files from being synchronized, by editing my `onedrive/config` to contain this:
```
skip_file = ".lock"
```
If I'm just using the config wrong, please let me know, but as far as I can tell, it should be blocking `.lock` files, and it's still warning me about them. If I need to, I'm willing to disable all warnings; though I would rather have them on just in case, I can't find a way to do that either.

Thanks in advance!

**Comment by abraunegg** [comment]
@gusruben 

The issue you are facing stems from how the application processes files and in what order. The order is:
1. Is the file a valid item
2. Does the file align to the Microsoft conventions (name, UTF sequence, unique name (based on Microsoft's interpretation ..))
3. Does the user want to include|exclude this file ('skip_file', 'skip_dir', 'sync_list' etc)

As `.lock` is an invalid file based on the [Microsoft requirements](https://support.microsoft.com/en-us/office/restrictions-and-limitations-in-onedrive-and-sharepoint-64883a5d-228e-48f5-b3d2-eb39e07630fa#invalidfilefoldernames), the application configuration 'skip_file' has no effect here (as it is processed later) - so you will always hit the first validation and get the skip message you are seeing:
```
Skipping item - invalid name (Microsoft Naming Convention): ./lock_files/.lock
```

Now .. lets say these lock files were 'skippable' ... you would still get a skipped item being logged:
```
Skipping item - excluded by skip_file config: ./lock_files/.lock2
```
Admittedly, this skipped item notification would only be displayed if using `--verbose` ...

The only options for resolution here are:
1. Clean up your OneDrive folder so that `.lock` files dont exist ... they really shouldn't be there, so you probably need to ask yourself why are they there in the first place ...
2. Ignore this and filter this out manually when using the application.

Changing the application code here so that when a file is being skipped due to a Microsoft Naming Convention is only displayed when using `--verbose` is not a viable proposition ... as you then would have no way of knowing if that important file *you think* is OK - but is not uploaded ... and you have no idea why .. which may lead to a data loss scenario if you then delete it locally.

**Reply by newlukai** [reply]
I'm also encountering this issue. I've a `.lock` file in the directory to synchronize and a KDE notification about that skipped item pops up every 5 minutes. That is annoying.

I understand _why_ this message appears, but I don't understand why we couldn't tackle this issue an some way.

In my setup I'm using the Linux version of the Joplin note taking app. While that app also allows to synchronize with OneDrive I prefer to let that app write its files to disk and use this `onedrive` client to synchronize with OneDrive. Unfortunately that app writes a `.lock` file.

So I see these options right now:
- Let Joplin synchronize with OneDrive on its own.
- Let Joplin write to a different directory and run a cron job to copy every file except the `.lock` file to a `onedrive` synchronized directory.
- Ask or contribute to Joplin to rename that file.
- Ask or contribute to `onedrive` to switch its file processing order.

I prefer the last option, because `onedrive` is a Linux tool. Microsoft file name restrictions do not apply on Linux machines, so it's perfectly fine for those environments to have these files. Of course it's not possible to synchronize these and the user should be noticed about files which are skipped. But in my opinion it should be possible for Linux users to tell `onedrive` that a notification about skipping a file is not necessary.

**Comment by abraunegg** [comment, marked as answer]
@gusruben , @newlukai

Please can you test the following PR to potentially your issue:

```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/2238/head:pr2238
git checkout pr2238
```

With this PR, the testing for Microsoft compliance is done post checking for user skipped items. Any item in 'skip_file|skip_dir' does not trigger a GUI warning.

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any options needed>
```

To install the PR, you will need to perform `sudo make install` to install the PR version. 

When running the PR, your version should be: `onedrive v2.4.21-30-g36c4d51` or greater.

**Reply by newlukai** [reply]
Thanks for that chance, works like a charm :)

Means, I just kept it running for half an hour and created/deleted files arbitrarily on both sides and got no warning regarding the skipped files. I must admit that I already had the corresponding `skip_file` in the config and did not test if the warning appears again once I remove that entry.

**Reply by abraunegg** [reply]
Thanks for the confirmation - will merge this PR into 'master'. 

If you do remove the 'skip_file' entry, the warning will appear again.
